### PR TITLE
Sha1

### PR DESCRIPTION
**Pull request by ilevyor**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by justinboswell — CHANGES_REQUESTED** [review]
Let's come back to this review once you've removed the non-sha1 code

File: source/darwin/commoncrypto_sha.c [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    struct cc_sha256_hash *ctx = hash->impl;
+    aws_mem_release(hash->allocator, ctx);
+}
+
+static int s_update_common(
+    struct aws_hash *hash,
+    void *cc_hash,
+    const struct aws_byte_cursor *to_hash,
+    int (*cc_sha_update)(void *, const void *, CC_LONG)) {
+    if (!hash->good) {
```
Comment (graebm) [inline source/darwin/commoncrypto_sha.c:105]: trivial styling:
1) rename ~~cc_sha_update~~ -> cc_sha_update_fn
2) put a newline between the end of the function declaration and the body. the declaration wrapped so it wasn't clear where it ends the the body begins

File: source/darwin/commoncrypto_sha.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+
+    cc_sha_update(cc_hash, to_hash->ptr, (CC_LONG)to_hash->len);
```
Comment (graebm) [inline source/darwin/commoncrypto_sha.c:109]: I'd just have 2 different files. The way these share code makes it more confusing than useful

File: source/hash.c [hunk 3]
```diff
@@ -5,10 +5,13 @@
 #include <aws/cal/hash.h>
 
 #ifndef BYO_CRYPTO
+// how does the linker know which default to link to?
```
Comment (graebm) [inline source/hash.c:8]: take this out

File: source/unix/opensslcrypto_hash.c [hunk 4]
```diff
@@ -99,6 +107,35 @@ struct aws_hash *aws_sha256_default_new(struct aws_allocator *allocator) {
...
+    hash->good = true;
+
+    if (!hash->impl) {
+        s_destroy(hash);
+        aws_raise_error(AWS_ERROR_OOM);
+        return NULL;
+    }
+
+    if (!g_aws_openssl_evp_md_ctx_table->init_ex_fn(ctx, EVP_sha1(), NULL)) {
+        s_destroy(hash);
```
Comment (graebm) [inline source/unix/opensslcrypto_hash.c:131]: we need to clean up whatever was created by `EVP_MD_CTX_new()`.
I can't focus enough to figure out which function pointer pointer pointer that is

fix this in the other implementations as well

File: source/windows/bcrypt_hash.c [hunk 5]
```diff
@@ -92,6 +114,35 @@ struct aws_hash *aws_sha256_default_new(struct aws_allocator *allocator) {
     return &bcrypt_hash->hash;
 }
 
+struct aws_hash *aws_sha1_default_new(struct aws_allocator *allocator) {
+    aws_thread_call_once(&s_sha1_once, s_load_sha1_alg_handle, NULL);
+
+    struct bcrypt_hash_handle *bcrypt_hash;
```
Comment (rccarper) [inline source/windows/bcrypt_hash.c:120]: Initialize both of these to NULL. If aws_mem_acquire_many were to fail, these could be their initial random values. (https://github.com/awslabs/aws-c-common/blob/main/source/allocator.c#L190).

File: source/windows/bcrypt_hash.c [hunk 6]
```diff
@@ -92,6 +114,35 @@ struct aws_hash *aws_sha256_default_new(struct aws_allocator *allocator) {
...
+    bcrypt_hash->hash.allocator = allocator;
+    bcrypt_hash->hash.vtable = &s_sha1_vtable;
+    bcrypt_hash->hash.impl = bcrypt_hash;
+    bcrypt_hash->hash.digest_size = AWS_SHA1_LEN;
+    bcrypt_hash->hash.good = true;
+    bcrypt_hash->hash_obj = hash_obj;
+    NTSTATUS status = BCryptCreateHash(
+        s_sha1_alg, &bcrypt_hash->hash_handle, bcrypt_hash->hash_obj, (ULONG)s_sha1_obj_len, NULL, 0, 0);
+
+    if (((NTSTATUS)status) < 0) {
```
Comment (rccarper) [inline source/windows/bcrypt_hash.c:138]: Trivial: cast to NTSTATUS looks redundant?

File: tests/sha1_test.c [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+AWS_TEST_CASE(sha1_nist_test_case_5, s_sha1_nist_test_case_5_fn)
+
+static int s_sha1_nist_test_case_5_truncated_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input = aws_byte_cursor_from_c_str("a");
```
Comment (rccarper) [inline tests/sha1_test.c:119]: Trivial: could also use AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL.

File: tests/sha1_test.c [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input =
+        aws_byte_cursor_from_c_str("abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno");
+
+    for (size_t i = 0; i < 16777216; ++i) {
```
Comment (rccarper) [inline tests/sha1_test.c:154]: Same comment as above--is 16777216 significant?

File: tests/sha1_test.c [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+static int s_sha1_nist_test_case_5_truncated_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input = aws_byte_cursor_from_c_str("a");
+
+    for (size_t i = 0; i < 1000000; ++i) {
```
Comment (rccarper) [inline tests/sha1_test.c:121]: Is 1000000 significant? Seems like it should be a local constant with maybe a comment.

File: tests/sha1_test.c [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_byte_cursor input = aws_byte_cursor_from_c_str("abcdefghbcdefghicdefghijdefghijkefghijklfghij"
+                                                              "klmghijklmnhijklmnoijklmnopjklmnopqklm"
+                                                              "nopqrlmnopqrsmnopqrstnopqrstu");
+    uint8_t output[AWS_SHA1_LEN] = {0};
+    struct aws_byte_buf output_buf = aws_byte_buf_from_array(output, sizeof(output));
+    output_buf.len = 1;
```
Comment (rccarper) [inline tests/sha1_test.c:190]: Why not just generate the byte buffer from a one character array? :)

File: include/aws/cal/hash.h [hunk 11]
```diff
@@ -84,6 +90,20 @@ AWS_CAL_API int aws_sha256_compute(
...
     size_t truncate_to);
 
+/**
+ * Computes the sha1 hash over input and writes the digest output to 'output'.
+ * Use this if you don't need to stream the data you're hashing and you can load
+ * the entire input to hash into memory. If you specify truncate_to to something
+ * other than 0, the output will be truncated to that  number of bytes. For
+ * example if you want a SHA1 digest as the first 16 bytes, set truncate_to
+ * to 16. If you want the full digest size, just set this to 0.
+ */
```
Comment (rccarper) [inline include/aws/cal/hash.h:100]: Trivial: single space between 'that' and 'number', and then comma after "For example".

File: source/unix/opensslcrypto_hash.c (hunk 4 above)
Comment (graebm) [reply source/unix/opensslcrypto_hash.c:131]: nevermind. the s_destroy() function handles it as you've pointed out

File: tests/sha1_test.c (hunk 7 above)
Comment (graebm) [reply tests/sha1_test.c:119]: counterpoint: it's a test, and this is actually more readable

Comment (graebm) [reply tests/sha1_test.c:119]: Eh. I think this is fine
reasoning:
- the macro avoids the O(N) runtime cost, but this is a test, so do whatever is more readable
- this function is more readable than the very-long-named macro

File: tests/sha1_test.c (hunk 8 above)
Comment (graebm) [reply tests/sha1_test.c:154]: Ryan, This test is named "nist_test_case_1", so I think these values come from some official documentation.

Itamar, consider putting a link to the definition for this test, or have a comment with an exact term that the reader could google to find it, or copy/paste the definition of the test from whatever doc it's coming from. The idea is: if this test fails in some far-flung future. Some poor engineer is going to need to figure out exactly what it was trying to do, so you need to give them the best possible clues

**Review by rccarper — APPROVED** [review]

File: builder.json [hunk 12]
```diff
@@ -1,20 +1,24 @@
 {
     "name": "aws-c-cal",
     "upstream": [
-        { "name": "aws-c-common" }
-    ],
-    "downstream": [
-
+        {
```
Comment (justinboswell) [inline builder.json:4]: This file has functionally no changes, could revert it.

File: source/darwin/commoncrypto_sha256.c [hunk 13]
```diff
@@ -2,6 +2,7 @@
  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0.
  */
+
```
Comment (justinboswell) [inline source/darwin/commoncrypto_sha256.c:5]: undo this change

**Review by justinboswell — APPROVED** [review]
Fix-n-ship!

**Review by justinboswell — APPROVED** [review]

**Review by graebm — APPROVED** [review]
